### PR TITLE
Add initial support for MidnightBSD

### PR DESCRIPTION
**Pull request by laffer1**
This build works on MidnightBSD 3.2.1+ (stable/3.2 branch or current)

### CODE REVIEW

**Comment by CarterLi** [comment]
Does GCC on MidnightBSD define `__FreeBSD__`? Does this patch work (as what I did for DragonFly)?

```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index cc33bf5b..e3481ec0 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -25,6 +25,9 @@ elseif("${CMAKE_SYSTEM_NAME}" STREQUAL "NetBSD")
 elseif("${CMAKE_SYSTEM_NAME}" STREQUAL "DragonFly")
     set(FreeBSD TRUE CACHE BOOL "..." FORCE)
     set(DragonFly TRUE CACHE BOOL "..." FORCE)
+elseif("${CMAKE_SYSTEM_NAME}" STREQUAL "MidnightBSD")
+    set(FreeBSD TRUE CACHE BOOL "..." FORCE)
+    set(MidnightBSD TRUE CACHE BOOL "..." FORCE)
 elseif("${CMAKE_SYSTEM_NAME}" STREQUAL "SunOS")
     set(SunOS TRUE CACHE BOOL "..." FORCE)
 elseif(NOT APPLE AND NOT WIN32)
```

**Comment by CarterLi** [comment]
Does GCC define `__MidnightBSD__` or something similar?

File: CMakeLists.txt [hunk 1]
```diff
@@ -51,36 +53,36 @@ include(CheckIncludeFile)
...
 cmake_dependent_option(ENABLE_FREETYPE "Enable freetype" ON "ANDROID" OFF)
-cmake_dependent_option(ENABLE_PULSE "Enable pulse" ON "LINUX OR OpenBSD OR NetBSD OR SunOS" OFF)
+cmake_dependent_option(ENABLE_PULSE "Enable pulse" ON "LINUX OR OpenBSD OR MidnightBSD OR NetBSD OR SunOS" OFF)
 cmake_dependent_option(ENABLE_DDCUTIL "Enable ddcutil" ON "LINUX" OFF)
 cmake_dependent_option(ENABLE_DIRECTX_HEADERS "Enable DirectX headers for WSL" ON "LINUX" OFF)
 cmake_dependent_option(ENABLE_ELF "Enable libelf" ON "LINUX OR ANDROID OR DragonFly" OFF)
 cmake_dependent_option(ENABLE_THREADS "Enable multithreading" ON "Threads_FOUND" OFF)
 cmake_dependent_option(ENABLE_LIBZFS "Enable libzfs" ON "LINUX OR FreeBSD OR SunOS" OFF)
-cmake_dependent_option(ENABLE_PCIACCESS "Enable libpciaccess" ON "NetBSD OR OpenBSD OR SunOS" OFF)
+cmake_dependent_option(ENABLE_PCIACCESS "Enable libpciaccess" ON "NetBSD OR OpenBSD OR MidnightBSD OR SunOS" OFF)
```
Comment (CarterLi) [inline CMakeLists.txt:85]: On FreeBSD, we use `ioctl` to talk [`/dev/pci`](https://man.freebsd.org/cgi/man.cgi?pci(4)) directly, which supports GPU driver detection but does not require external libs to be installed. Doesn't MidnightBSD have that?

File: CMakeLists.txt [hunk 2]
```diff
@@ -51,36 +53,36 @@ include(CheckIncludeFile)
...
-cmake_dependent_option(ENABLE_GLX "Enable glx" ON "LINUX OR FreeBSD OR OpenBSD OR NetBSD OR SunOS" OFF)
-cmake_dependent_option(ENABLE_OSMESA "Enable osmesa" ON "LINUX OR FreeBSD OR OpenBSD OR NetBSD OR SunOS" OFF)
-cmake_dependent_option(ENABLE_OPENCL "Enable opencl" ON "LINUX OR FreeBSD OR OpenBSD OR NetBSD OR WIN32 OR ANDROID OR SunOS" OFF)
+cmake_dependent_option(ENABLE_EGL "Enable egl" ON "LINUX OR FreeBSD OR OpenBSD OR MidnightBSD OR NetBSD OR WIN32 OR SunOS" OFF)
+cmake_dependent_option(ENABLE_GLX "Enable glx" ON "LINUX OR FreeBSD OR OpenBSD OR MidnightBSD OR NetBSD OR SunOS" OFF)
+cmake_dependent_option(ENABLE_OSMESA "Enable osmesa" ON "LINUX OR FreeBSD OR OpenBSD OR MidnightBSD OR NetBSD OR SunOS" OFF)
+cmake_dependent_option(ENABLE_OPENCL "Enable opencl" ON "LINUX OR FreeBSD OR OpenBSD OR MidnightBSD OR NetBSD OR WIN32 OR ANDROID OR SunOS" OFF)
 cmake_dependent_option(ENABLE_FREETYPE "Enable freetype" ON "ANDROID" OFF)
-cmake_dependent_option(ENABLE_PULSE "Enable pulse" ON "LINUX OR OpenBSD OR NetBSD OR SunOS" OFF)
+cmake_dependent_option(ENABLE_PULSE "Enable pulse" ON "LINUX OR OpenBSD OR MidnightBSD OR NetBSD OR SunOS" OFF)
```
Comment (CarterLi) [inline CMakeLists.txt:79]: This is for sound device detection. On FreeBSD we use `ioctl` to grab info from `/dev/mixer*` directly. If it's supported by `MidnightOS` pulseaudio is not needed. In addition, to use pulseaudio, `src/detection/sound/sound_linux.c` must be appended in `LIBFASTFETCH_SRC`

File: CMakeLists.txt (hunk 1 above)
Comment (CarterLi) [reply CMakeLists.txt:85]: Similar to libpulse, `src/detection/gpu/gpu_general.c` must be appended in `LIBFASTFETCH_SRC` in order to let fastfetch use `libpciaccess`

File: CMakeLists.txt [hunk 3]
```diff
@@ -1123,7 +1125,7 @@ if(APPLE AND EXISTS "/usr/bin/otool")
         target_compile_definitions(libfastfetch PUBLIC FF_LIBSYSTEM_VERSION="${CMAKE_MATCH_1}")
     endif()
 endif()
-if(FreeBSD AND EXISTS "/usr/local/bin/objdump")
+if((FreeBSD OR MidnightBSD) AND EXISTS "/usr/local/bin/objdump")
     execute_process(COMMAND /bin/sh -c "/usr/local/bin/objdump -T /lib/libc.so.* | grep 'FBSD_[0-9][0-9]*\\.[0-9][0-9]*' -o | sort -Vru | head -1"
```
Comment (CarterLi) [inline CMakeLists.txt:1129]: Does MidnightBSD use FreeBSD libc?

Comment (laffer1) [reply CMakeLists.txt:1129]: MidnightBSD uses FreeBSD libc with some additions.  So there are FBSD symbol versions and also MBSD versions.

**Comment by laffer1** [comment]
MidnightBSD defines both __MidnightBSD__ and __FreeBSD__ in current versions.

File: CMakeLists.txt (hunk 1 above)
Comment (laffer1) [reply CMakeLists.txt]: good catch. fixed.

File: CMakeLists.txt (hunk 3 above)
Comment (CarterLi) [reply CMakeLists.txt:1130]: I guess using MBSD versions is better.

**Comment by CarterLi** [comment]
```diff
diff --git a/src/detection/version/version.c b/src/detection/version/version.c
index 8b230eec..1f1deb34 100644
--- a/src/detection/version/version.c
+++ b/src/detection/version/version.c
@@ -28,6 +28,8 @@
     #define FF_SYSNAME "Linux"
 #elif defined(__DragonFly__) // We define `__FreeBSD__` on DragonFly BSD for simplification
     #define FF_SYSNAME "DragonFly"
+#elif defined(__MidnightBSD__)
+    #define FF_SYSNAME "MidnightBSD"
 #elif defined(__FreeBSD__)
     #define FF_SYSNAME "FreeBSD"
 #elif defined(__APPLE__)
```

**Pull request merged**